src/tools/rbd/action/BenchWrite.cc

```diff
@@ -262,6 +262,11 @@ int execute(const po::variables_map &vm) {
   } else {
     bench_io_size = 4096;
   }
+  if ((bench_io_size > 100*1024*1024) || (bench_io_size < 1))
```

Comment:
@dillaman 


root@sys0:~# rbd bench-write foo --io-size 1G 
*** Caught signal (Floating point exception) **
 in thread 7ff11a7ef4c0
 ceph version 9.0.2-5533-gf3e58a0 (f3e58a054aed197bf4f872c0af7b003f80e0ea2f)
 1: (()+0x28cb3a) [0x7ff11a54eb3a]
 2: (()+0x10340) [0x7ff106493340]
 3: (rbd::action::bench_write::do_bench_write(librbd::Image&, unsigned long, unsigned long, unsigned long, bool)+0x3af) [0x7ff11a400c0f]
 4: (rbd::action::bench_write::execute(boost::program_options::variables_map const&)+0x382) [0x7ff11a402452]
 5: (rbd::Shell::execute(int, char const**)+0x7cb) [0x7ff11a3f23cb]
 6: (main()+0x62) [0x7ff11a3c55f2]
 7: (__libc_start_main()+0xf5) [0x7ff1049c7f45]
 8: (()+0x11ada7) [0x7ff11a3dcda7]
2016-05-25 22:10:08.941332 7ff11a7ef4c0 -1 *** Caught signal (Floating point exception) **
 in thread 7ff11a7ef4c0

 ceph version 9.0.2-5533-gf3e58a0 (f3e58a054aed197bf4f872c0af7b003f80e0ea2f)
 1: (()+0x28cb3a) [0x7ff11a54eb3a]
 2: (()+0x10340) [0x7ff106493340]
 3: (rbd::action::bench_write::do_bench_write(librbd::Image&, unsigned long, unsigned long, unsigned long, bool)+0x3af) [0x7ff11a400c0f]
 4: (rbd::action::bench_write::execute(boost::program_options::variables_map const&)+0x382) [0x7ff11a402452]
 5: (rbd::Shell::execute(int, char const**)+0x7cb) [0x7ff11a3f23cb]
 6: (main()+0x62) [0x7ff11a3c55f2]
 7: (__libc_start_main()+0xf5) [0x7ff1049c7f45]
 8: (()+0x11ada7) [0x7ff11a3dcda7]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.

--- begin dump of recent events ---
   -80> 2016-05-25 22:10:08.552412 7ff11a7ef4c0  5 asok(0x7ff1255470a0) register_command perfcounters_dump hook 0x7ff12554b060
   -79> 2016-05-25 22:10:08.552434 7ff11a7ef4c0  5 asok(0x7ff1255470a0) register_command 1 hook 0x7ff12554b060
   -78> 2016-05-25 22:10:08.552437 7ff11a7ef4c0  5 asok(0x7ff1255470a0) register_command perf dump hook 0x7ff12554b060
   -77> 2016-05-25 22:10:08.552440 7ff11a7ef4c0  5 asok(0x7ff1255470a0) register_command perfcounters_schema hook 0x7ff12554b060
   -76> 2016-05-25 22:10:08.552442 7ff11a7ef4c0  5 asok(0x7ff1255470a0) register_command 2 hook 0x7ff12554b060
   -75> 2016-05-25 22:10:08.552444 7ff11a7ef4c0  5 asok(0x7ff1255470a0) register_command perf schema hook 0x7ff12554b060
   -74> 2016-05-25 22:10:08.552448 7ff11a7ef4c0  5 asok(0x7ff1255470a0) register_command perf reset hook 0x7ff12554b060
   -73> 2016-05-25 22:10:08.552455 7ff11a7ef4c0  5 asok(0x7ff1255470a0) register_command config show hook 0x7ff12554b060
   -72> 2016-05-25 22:10:08.552460 7ff11a7ef4c0  5 asok(0x7ff1255470a0) register_command config set hook 0x7ff12554b060
   -71> 2016-05-25 22:10:08.552462 7ff11a7ef4c0  5 asok(0x7ff1255470a0) register_command config get hook 0x7ff12554b060
   -70> 2016-05-25 22:10:08.552464 7ff11a7ef4c0  5 asok(0x7ff1255470a0) register_command config diff hook 0x7ff12554b060
   -69> 2016-05-25 22:10:08.552468 7ff11a7ef4c0  5 asok(0x7ff1255470a0) register_command log flush hook 0x7ff12554b060
   -68> 2016-05-25 22:10:08.552472 7ff11a7ef4c0  5 asok(0x7ff1255470a0) register_command log dump hook 0x7ff12554b060
   -67> 2016-05-25 22:10:08.552479 7ff11a7ef4c0  5 asok(0x7ff1255470a0) register_command log reopen hook 0x7ff12554b060
   -66> 2016-05-25 22:10:08.559473 7ff11a7ef4c0 10 monclient(hunting): build_initial_monmap
   -65> 2016-05-25 22:10:08.559541 7ff11a7ef4c0  1 librados: starting msgr at :/0
   -64> 2016-05-25 22:10:08.559547 7ff11a7ef4c0  1 librados: starting objecter
   -63> 2016-05-25 22:10:08.559627 7ff11a7ef4c0  5 asok(0x7ff1255470a0) register_command objecter_requests hook 0x7ff1255a24f0
   -62> 2016-05-25 22:10:08.559745 7ff11a7ef4c0  1 -- :/0 messenger.start
   -61> 2016-05-25 22:10:08.559790 7ff11a7ef4c0  1 librados: setting wanted keys
   -60> 2016-05-25 22:10:08.559794 7ff11a7ef4c0  1 librados: calling monclient init
   -59> 2016-05-25 22:10:08.559795 7ff11a7ef4c0 10 monclient(hunting): init
   -58> 2016-05-25 22:10:08.559807 7ff11a7ef4c0  5 adding auth protocol: cephx
   -57> 2016-05-25 22:10:08.559811 7ff11a7ef4c0 10 monclient(hunting): auth_supported 2 method cephx
   -56> 2016-05-25 22:10:08.559970 7ff11a7ef4c0  2 auth: KeyRing::load: loaded key file /etc/ceph/ceph.client.admin.keyring
   -55> 2016-05-25 22:10:08.560079 7ff11a7ef4c0 10 monclient(hunting): _reopen_session rank -1 name 
   -54> 2016-05-25 22:10:08.560150 7ff11a7ef4c0 10 monclient(hunting): picked mon.noname-a con 0x7ff1255a84f0 addr 172.16.41.70:6789/0
   -53> 2016-05-25 22:10:08.560167 7ff11a7ef4c0 10 monclient(hunting): _send_mon_message to mon.noname-a at 172.16.41.70:6789/0
   -52> 2016-05-25 22:10:08.560172 7ff11a7ef4c0  1 -- :/569358439 --> 172.16.41.70:6789/0 -- auth(proto 0 30 bytes epoch 0) v1 -- ?+0 0x7ff1255a9a40 con 0x7ff1255a84f0
   -51> 2016-05-25 22:10:08.560185 7ff11a7ef4c0 10 monclient(hunting): renew_subs
   -50> 2016-05-25 22:10:08.560188 7ff11a7ef4c0 10 monclient(hunting): authenticate will time out at 2016-05-25 22:15:08.560188
   -49> 2016-05-25 22:10:08.560592 7ff11a7e7700  1 -- 172.16.41.70:0/569358439 learned my addr 172.16.41.70:0/569358439
   -48> 2016-05-25 22:10:08.560977 7ff11a2be700  2 -- 172.16.41.70:0/569358439 >> 172.16.41.70:6789/0 pipe(0x7ff1255ab550 sd=3 :58501 s=2 pgs=77965 cs=1 l=1 c=0x7ff1255a84f0).reader got KEEPALIVE_ACK
   -47> 2016-05-25 22:10:08.561279 7ff0ff332700  1 -- 172.16.41.70:0/569358439 <== mon.0 172.16.41.70:6789/0 1 ==== mon_map magic: 0 v1 ==== 482+0+0 (162053117 0 0) 0x7ff0f0000d90 con 0x7ff1255a84f0
   -46> 2016-05-25 22:10:08.561314 7ff0ff332700 10 monclient(hunting): handle_monmap mon_map magic: 0 v1
   -45> 2016-05-25 22:10:08.561336 7ff0ff332700 10 monclient(hunting):  got monmap 2, mon.noname-a is now rank -1
   -44> 2016-05-25 22:10:08.561339 7ff0ff332700 10 monclient(hunting): dump:
epoch 2
fsid 2b1cf291-5332-4d6a-90ba-d4ce3fd0195a
last_changed 2016-05-19 04:05:42.720685
created 0.000000
0: 172.16.41.70:6789/0 mon.sys0
1: 172.16.41.71:6789/0 mon.sys1
2: 172.16.41.72:6789/0 mon.sys2

   -43> 2016-05-25 22:10:08.561371 7ff0ff332700  1 -- 172.16.41.70:0/569358439 <== mon.0 172.16.41.70:6789/0 2 ==== auth_reply(proto 2 0 (0) Success) v1 ==== 33+0+0 (78004728 0 0) 0x7ff0f00012f0 con 0x7ff1255a84f0
   -42> 2016-05-25 22:10:08.561412 7ff0ff332700 10 monclient(hunting): my global_id is 4305
   -41> 2016-05-25 22:10:08.561545 7ff0ff332700 10 monclient(hunting): _send_mon_message to mon.sys0 at 172.16.41.70:6789/0
   -40> 2016-05-25 22:10:08.561558 7ff0ff332700  1 -- 172.16.41.70:0/569358439 --> 172.16.41.70:6789/0 -- auth(proto 2 32 bytes epoch 0) v1 -- ?+0 0x7ff0f4001730 con 0x7ff1255a84f0
   -39> 2016-05-25 22:10:08.562208 7ff0ff332700  1 -- 172.16.41.70:0/569358439 <== mon.0 172.16.41.70:6789/0 3 ==== auth_reply(proto 2 0 (0) Success) v1 ==== 206+0+0 (1237060503 0 0) 0x7ff0f00012f0 con 0x7ff1255a84f0
   -38> 2016-05-25 22:10:08.562349 7ff0ff332700 10 monclient(hunting): _send_mon_message to mon.sys0 at 172.16.41.70:6789/0
   -37> 2016-05-25 22:10:08.562361 7ff0ff332700  1 -- 172.16.41.70:0/569358439 --> 172.16.41.70:6789/0 -- auth(proto 2 165 bytes epoch 0) v1 -- ?+0 0x7ff0f4006a60 con 0x7ff1255a84f0
   -36> 2016-05-25 22:10:08.563309 7ff0ff332700  1 -- 172.16.41.70:0/569358439 <== mon.0 172.16.41.70:6789/0 4 ==== auth_reply(proto 2 0 (0) Success) v1 ==== 393+0+0 (710160174 0 0) 0x7ff0f00012f0 con 0x7ff1255a84f0
   -35> 2016-05-25 22:10:08.563432 7ff0ff332700  1 monclient(hunting): found mon.sys0
   -34> 2016-05-25 22:10:08.563439 7ff0ff332700 10 monclient: _send_mon_message to mon.sys0 at 172.16.41.70:6789/0
   -33> 2016-05-25 22:10:08.563447 7ff0ff332700  1 -- 172.16.41.70:0/569358439 --> 172.16.41.70:6789/0 -- mon_subscribe({monmap=0+}) v2 -- ?+0 0x7ff1255a6010 con 0x7ff1255a84f0
   -32> 2016-05-25 22:10:08.563486 7ff11a7ef4c0  5 monclient: authenticate success, global_id 4305
   -31> 2016-05-25 22:10:08.563514 7ff11a7ef4c0 10 monclient: renew_subs
   -30> 2016-05-25 22:10:08.563521 7ff11a7ef4c0 10 monclient: _send_mon_message to mon.sys0 at 172.16.41.70:6789/0
   -29> 2016-05-25 22:10:08.563529 7ff11a7ef4c0  1 -- 172.16.41.70:0/569358439 --> 172.16.41.70:6789/0 -- mon_subscribe({osdmap=0}) v2 -- ?+0 0x7ff1255a9a40 con 0x7ff1255a84f0
   -28> 2016-05-25 22:10:08.563607 7ff11a7ef4c0 10 monclient: renew_subs - empty
   -27> 2016-05-25 22:10:08.563799 7ff0ff332700  1 -- 172.16.41.70:0/569358439 <== mon.0 172.16.41.70:6789/0 5 ==== mon_map magic: 0 v1 ==== 482+0+0 (162053117 0 0) 0x7ff0f0001610 con 0x7ff1255a84f0
   -26> 2016-05-25 22:10:08.563818 7ff0ff332700 10 monclient: handle_monmap mon_map magic: 0 v1
   -25> 2016-05-25 22:10:08.563834 7ff0ff332700 10 monclient:  got monmap 2, mon.sys0 is now rank 0
   -24> 2016-05-25 22:10:08.563839 7ff0ff332700 10 monclient: dump:
epoch 2
fsid 2b1cf291-5332-4d6a-90ba-d4ce3fd0195a
last_changed 2016-05-19 04:05:42.720685
created 0.000000
0: 172.16.41.70:6789/0 mon.sys0
1: 172.16.41.71:6789/0 mon.sys1
2: 172.16.41.72:6789/0 mon.sys2

   -23> 2016-05-25 22:10:08.563914 7ff0ff332700  1 -- 172.16.41.70:0/569358439 <== mon.0 172.16.41.70:6789/0 6 ==== osd_map(95..95 src has 1..95) v3 ==== 12667+0+0 (1983691057 0 0) 0x7ff0f00012f0 con 0x7ff1255a84f0
   -22> 2016-05-25 22:10:08.564348 7ff11a7ef4c0  1 librados: init done
   -21> 2016-05-25 22:10:08.564441 7ff11a7ef4c0  5 asok(0x7ff1255470a0) register_command rbd cache flush rbd/foo hook 0x7ff1255a53f0
   -20> 2016-05-25 22:10:08.564456 7ff11a7ef4c0  5 asok(0x7ff1255470a0) register_command rbd cache invalidate rbd/foo hook 0x7ff1255a53f0
   -19> 2016-05-25 22:10:08.564527 7ff11a7ef4c0  5 librbd::AioImageRequestWQ: 0x7ff1255a5f90 : ictx=0x7ff1255a7440
   -18> 2016-05-25 22:10:08.564748 7ff11a7ef4c0  1 -- 172.16.41.70:0/569358439 --> 172.16.41.70:6816/2581 -- osd_op(client.4305.0:1 0.6a181655 rbd_id.foo [stat] snapc 0=[] ack+read+known_if_redirected e95) v7 -- ?+0 0x7ff1255b7040 con 0x7ff1255b5c60
   -17> 2016-05-25 22:10:08.566473 7ff0fc22b700  1 -- 172.16.41.70:0/569358439 <== osd.4 172.16.41.70:6816/2581 1 ==== osd_op_reply(1 rbd_id.foo [stat] v0'0 uv2 ondisk = 0) v6 ==== 177+0+16 (1007905865 0 1522554074) 0x7ff0e4000cb0 con 0x7ff1255b5c60
   -16> 2016-05-25 22:10:08.566640 7ff0effff700  1 -- 172.16.41.70:0/569358439 --> 172.16.41.70:6816/2581 -- osd_op(client.4305.0:2 0.6a181655 rbd_id.foo [call rbd.get_id] snapc 0=[] ack+read+known_if_redirected e95) v7 -- ?+0 0x7ff0d8002250 con 0x7ff1255b5c60
   -15> 2016-05-25 22:10:08.567637 7ff0fc22b700  1 -- 172.16.41.70:0/569358439 <== osd.4 172.16.41.70:6816/2581 2 ==== osd_op_reply(2 rbd_id.foo [call] v0'0 uv2 ondisk = 0) v6 ==== 177+0+16 (2752176729 0 2941236676) 0x7ff0e4000cb0 con 0x7ff1255b5c60
   -14> 2016-05-25 22:10:08.567880 7ff0effff700  1 -- 172.16.41.70:0/569358439 --> 172.16.41.70:6820/3265 -- osd_op(client.4305.0:3 0.e096734e rbd_header.378b2facdf9b [call rbd.get_size,call rbd.get_object_prefix] snapc 0=[] ack+read+known_if_redirected e95) v7 -- ?+0 0x7ff0d8002030 con 0x7ff0d80046e0
   -13> 2016-05-25 22:10:08.569472 7ff0eeffd700  1 -- 172.16.41.70:0/569358439 <== osd.5 172.16.41.70:6820/3265 1 ==== osd_op_reply(3 rbd_header.378b2facdf9b [call,call] v0'0 uv126 ondisk = 0) v6 ==== 232+0+34 (479650515 0 440693653) 0x7ff0d0000cf0 con 0x7ff0d80046e0
   -12> 2016-05-25 22:10:08.569623 7ff0effff700  1 -- 172.16.41.70:0/569358439 --> 172.16.41.70:6820/3265 -- osd_op(client.4305.0:4 0.e096734e rbd_header.378b2facdf9b [call rbd.get_stripe_unit_count] snapc 0=[] ack+read+known_if_redirected e95) v7 -- ?+0 0x7ff0d8002030 con 0x7ff0d80046e0
   -11> 2016-05-25 22:10:08.570334 7ff0eeffd700  1 -- 172.16.41.70:0/569358439 <== osd.5 172.16.41.70:6820/3265 2 ==== osd_op_reply(4 rbd_header.378b2facdf9b [call] v0'0 uv0 ondisk = -8 ((8) Exec format error)) v6 ==== 190+0+0 (3881539227 0 0) 0x7ff0d0000cf0 con 0x7ff0d80046e0
   -10> 2016-05-25 22:10:08.570580 7ff0effff700  1 -- 172.16.41.70:0/569358439 --> 172.16.41.70:6820/3265 -- osd_op(client.4305.0:5 0.e096734e rbd_header.378b2facdf9b [call rbd.metadata_list] snapc 0=[] ack+read+known_if_redirected e95) v7 -- ?+0 0x7ff0d800c740 con 0x7ff0d80046e0
    -9> 2016-05-25 22:10:08.571875 7ff0eeffd700  1 -- 172.16.41.70:0/569358439 <== osd.5 172.16.41.70:6820/3265 3 ==== osd_op_reply(5 rbd_header.378b2facdf9b [call] v0'0 uv126 ondisk = 0) v6 ==== 190+0+4 (3697117247 0 0) 0x7ff0d0001370 con 0x7ff0d80046e0
    -8> 2016-05-25 22:10:08.572542 7ff0ef7fe700  1 -- 172.16.41.70:0/569358439 --> 172.16.41.70:6820/3265 -- osd_op(client.4305.0:6 0.e096734e rbd_header.378b2facdf9b [watch watch cookie 140672661198960] snapc 0=[] ondisk+write+known_if_redirected e95) v7 -- ?+0 0x7ff0e8002330 con 0x7ff0d80046e0
    -7> 2016-05-25 22:10:08.574921 7ff0eeffd700  1 -- 172.16.41.70:0/569358439 <== osd.5 172.16.41.70:6820/3265 4 ==== osd_op_reply(6 rbd_header.378b2facdf9b [watch watch cookie 140672661198960] v95'154 uv126 ondisk = 0) v6 ==== 190+0+0 (1038839968 0 0) 0x7ff0d0001370 con 0x7ff0d80046e0
    -6> 2016-05-25 22:10:08.575144 7ff0ef7fe700  1 -- 172.16.41.70:0/569358439 --> 172.16.41.70:6820/3265 -- osd_op(client.4305.0:7 0.e096734e rbd_header.378b2facdf9b [call rbd.get_size,call rbd.get_features,call rbd.get_snapcontext,call rbd.get_parent,call lock.get_info] snapc 0=[] ack+read+known_if_redirected e95) v7 -- ?+0 0x7ff0e800aaf0 con 0x7ff0d80046e0
    -5> 2016-05-25 22:10:08.576008 7ff0eeffd700  1 -- 172.16.41.70:0/569358439 <== osd.5 172.16.41.70:6820/3265 5 ==== osd_op_reply(7 rbd_header.378b2facdf9b [call,call,call,call,call] v0'0 uv126 ondisk = 0) v6 ==== 358+0+88 (2779631686 0 3321974444) 0x7ff0d0001340 con 0x7ff0d80046e0
    -4> 2016-05-25 22:10:08.576146 7ff0effff700  1 -- 172.16.41.70:0/569358439 --> 172.16.41.70:6820/3265 -- osd_op(client.4305.0:8 0.e096734e rbd_header.378b2facdf9b [call rbd.get_flags,call rbd.get_flags] snapc 0=[] ack+read+known_if_redirected e95) v7 -- ?+0 0x7ff0d8007000 con 0x7ff0d80046e0
    -3> 2016-05-25 22:10:08.576776 7ff0eeffd700  1 -- 172.16.41.70:0/569358439 <== osd.5 172.16.41.70:6820/3265 6 ==== osd_op_reply(8 rbd_header.378b2facdf9b [call,call] v0'0 uv126 ondisk = 0) v6 ==== 232+0+16 (1815295250 0 0) 0x7ff0d0001340 con 0x7ff0d80046e0
    -2> 2016-05-25 22:10:08.576916 7ff0effff700  1 -- 172.16.41.70:0/569358439 --> 172.16.41.70:6820/3265 -- osd_op(client.4305.0:9 0.e096734e rbd_header.378b2facdf9b [call rbd.get_snapshot_name,call rbd.get_size,call rbd.get_parent,call rbd.get_protection_status] snapc 0=[] ack+read+known_if_redirected e95) v7 -- ?+0 0x7ff0d8014050 con 0x7ff0d80046e0
    -1> 2016-05-25 22:10:08.577710 7ff0eeffd700  1 -- 172.16.41.70:0/569358439 <== osd.5 172.16.41.70:6820/3265 7 ==== osd_op_reply(9 rbd_header.378b2facdf9b [call,call,call,call] v0'0 uv126 ondisk = 0) v6 ==== 316+0+45 (4268658287 0 870807602) 0x7ff0d0001340 con 0x7ff0d80046e0
     0> 2016-05-25 22:10:08.941332 7ff11a7ef4c0 -1 *** Caught signal (Floating point exception) **
 in thread 7ff11a7ef4c0

 ceph version 9.0.2-5533-gf3e58a0 (f3e58a054aed197bf4f872c0af7b003f80e0ea2f)
 1: (()+0x28cb3a) [0x7ff11a54eb3a]
 2: (()+0x10340) [0x7ff106493340]
 3: (rbd::action::bench_write::do_bench_write(librbd::Image&, unsigned long, unsigned long, unsigned long, bool)+0x3af) [0x7ff11a400c0f]
 4: (rbd::action::bench_write::execute(boost::program_options::variables_map const&)+0x382) [0x7ff11a402452]
 5: (rbd::Shell::execute(int, char const**)+0x7cb) [0x7ff11a3f23cb]
 6: (main()+0x62) [0x7ff11a3c55f2]
 7: (__libc_start_main()+0xf5) [0x7ff1049c7f45]
 8: (()+0x11ada7) [0x7ff11a3dcda7]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.

--- logging levels ---
   0/ 5 none
   0/ 1 lockdep
   0/ 1 context
   1/ 1 crush
   1/ 5 mds
   1/ 5 mds_balancer
   1/ 5 mds_locker
   1/ 5 mds_log
   1/ 5 mds_log_expire
   1/ 5 mds_migrator
   0/ 1 buffer
   0/ 1 timer
   0/ 1 filer
   0/ 1 striper
   0/ 1 objecter
   0/ 5 rados
   0/ 5 rbd
   0/ 5 rbd_replay
   0/ 5 journaler
   0/ 5 objectcacher
   0/ 5 client
   0/ 5 osd
   0/ 5 optracker
   0/ 5 objclass
   1/ 3 filestore
   1/ 3 keyvaluestore
   1/ 3 journal
   0/ 5 ms
   1/ 5 mon
   0/10 monc
   1/ 5 paxos
   0/ 5 tp
   1/ 5 auth
   1/ 5 crypto
   1/ 1 finisher
   1/ 5 heartbeatmap
   1/ 5 perfcounter
   1/ 5 rgw
   1/10 civetweb
   1/ 5 javaclient
   1/ 5 asok
   1/ 1 throttle
   0/ 0 refs
   1/ 5 xio
   1/ 5 compressor
   1/ 5 newstore
   1/ 5 bluestore
   1/ 5 bluefs
   1/ 3 bdev
   1/ 5 kstore
   4/ 5 rocksdb
   4/ 5 leveldb
  -2/-2 (syslog threshold)
  99/99 (stderr threshold)
  max_recent       500
  max_new         1000
  log_file 
--- end dump of recent events ---
Floating point exception (core dumped)